[CVariant] Replace rapidjson with nlohmann-json

sundermann opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
This pull request focuses on replacing the usage of RapidJSON with nlohmann-json across the codebase.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
RapidJSON has not had an update for almost 10 years now. It's still actively developed but it seems Tencent is just not interested in tagging a release. See https://github.com/Tencent/rapidjson/issues/2284#issuecomment-2156004291

As a sideeffect, this should also unbreak the webOS release builds on CI which currently is broken due to a mixup of different rapidjson versions from core and addons.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tests pass, macOS, webOS

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

sundermann committed:
[Android] Replace rapidjson usage with CVariant

sundermann committed:
[CVariant] Replace rapidjson with nlohmann-json

fuzzard commented:
I'm all for a replacement. I had started looking at https://github.com/ibireme/yyjson , but anything is fine by me.

fuzzard commented:
Fixup for the build side of things
https://github.com/fuzzard/xbmc/commit/e7b39c0d13523841e49cfb564a980c6af882d8c4

fuzzard commented:
https://github.com/fuzzard/xbmc/commit/6b48b83f75076b93818f156a1d49e5148c54efe5

seems to fix android. Not sure if theres a better way to handle it, i'll let the c++ experts chime in

sundermann commented:
@fuzzard Thanks for your fixes. I've applied them. Also moved parsing to the `sax_parser` interface which closer mimics what we had before.

I had removed all the rapidjson stuff from the linux install pages but haven't added nlohmann-json back in. Should we do that? It should fall back to an internal build if the library is not found.

fuzzard commented:
Definitely should. Linux should always prefer system libs as a default over any internal building we provide

sundermann commented:
I have updated the installation docs for linux

sundermann commented:
`CVariant` always uses 64 bit integers internally, so there's no need to cast it explicitly. I think this is ready for review

fuzzard commented on CMakeLists.txt:
```diff
@@ -115,7 +115,7 @@ dependent_option(ENABLE_INTERNAL_FLATBUFFERS "Enable internal flatbuffers?")
 dependent_option(ENABLE_INTERNAL_NFS "Enable internal libnfs?")
 dependent_option(ENABLE_INTERNAL_PCRE2 "Enable internal pcre2?")
-dependent_option(ENABLE_INTERNAL_RapidJSON "Enable internal rapidjson?")
+dependent_option(ENABLE_INTERNAL_NlohmannJSON "Enable internal nlohmannjson?")
```
I would make this all uppercase. It makes things easier internally to rely on that standard, and was changed at some point in time prior to my involvement to be the defacto

fuzzard commented on CMakeLists.txt:
```diff
@@ -240,7 +240,7 @@ set(required_deps ASS>=0.15.0
                   OpenSSL>=1.1.0
                   PCRE2
-                  RapidJSON>=1.0.2
+                  NlohmannJSON>=3.2.0
```
Query, is the 3.2 a hard requirement? Just confirming.

sundermann replied:
Yes, `sax_parse` is available from 3.2.0 which was released in 2018

fuzzard reviewed (approved):
Other than a couple minors. things look ok for me on the build side.

I'll leave others to comment on the actual c++ side, but a minor scroll through it looked reasonable to me. Theres a fair bit of use of CJSONVariantPaser in the codebase, but the tests seem to be passing, so in theory it should be ok i guess.

Tolriq commented:
Can some quick performance tests be done to confirm the impact is not too high? Specially large JSON output of a couple of MB on low end devices like rpi.

sundermann commented:
While nlohmann-json is slower than rapidJSON, we're not handling large JSON datasets here. The performance difference will be negligible, so I recommend prioritizing maintainability over marginal performance gains in this case.

Tolriq commented:
There was tons of work to speed up json, there is a lot of large data set as soon as we talk API. The webinterface, any remote app like Kore, or Yatse will request large JSON to have the library data to show. Even chunked we are still talking about MB of data.
There's speed and memory usage to take in account to choose the successor, check the history of the migration to rapidJSON and the work of montellese.

acidzab commented:
Should you consider simdjson as a replace for RapidJSON ?

fuzzard commented:
PRs for alternatives are welcome

sundermann commented:
The concern seems mostly focused around serialization. The nice thing about rapidjson is that it has an explicit [writer interface](https://rapidjson.org/classrapidjson_1_1_writer.html) meaning that we can directly convert a `CVariant` into a string without requiring an intermediate json object representation. I haven't seen something similar in any other popular json library and it's also not something that `simdjson` could solve. 

Talking about deserialization `simdjson` does not have an SAX interface for parsing either so it would require us to first parse the whole json into a document and then constructing a `CVariant` from it which is somewhat inefficient as it requires twice the memory and a iterating a second time over the intermediate json representation.

Speaking of performance generally, the current approch for serialization is recursive which is somewhat inefficient as well. It could be rewritten using a stack.

Tolriq commented:
Honestly i don't recall all the details everything was made a very very long time ago :) 
And the new libs and compiler have maybe / probably improved enough so that the performance impact won't be that high.
All i ask is that there's some quick benchmark done to ensure there's no insane performance regression for the API users like the web interface, Kore and Yatse. It was a contention point with a lot of improvements, users will certain complain if everything take 50% of more time now, or just no more works because large chunks use too much memory to serialize on low end devices.

neo1973 commented:
I measured the `POST http://localhost:8080/jsonrpc?TVShowCollection` call on my PC in release build. Content of the call:

```json
{"jsonrpc":"2.0","method":"VideoLibrary.GetTVShows","id":"1745076795448","params":{"properties":["title","art","playcount","lastplayed","dateadded","episode","rating","year","file","genre","watchedepisodes","cast","studio","mpaa","tag"],"limits":{"start":0},"sort":{"method":"title","order":"ascending","ignorearticle":true}}}
```

This produces a 1.52 MB JSON response with my DB. The numbers:

- nlohmann: ~190ms
- rapidjson: ~170ms

So yes, slower, but not by a lot.

sundermann commented:
I measured 420ms using nlohmann-json and 370ms using Rapidjson on macOS (Debug + ASan) using the call above. In Release configuration the performance difference was minimal 60ms vs 58ms.

For fun, I tried converting the serialization into an iterative solution but the performance was much worse (2000ms). Probably because the stack needs to track a lot more with an iterative approach.

neo1973 commented on xbmc/utils/JSONVariantWriter.cpp:
```diff
@@ -10,83 +10,84 @@
     for (CVariant::const_iterator_array itr = value.begin_array(); itr != value.end_array(); ++itr)
     {
-      if (!InternalWrite(writer, *itr))
+      nlohmann::json o;
+      if (!InternalWrite(o, *itr))
         return false;
+      writer.emplace_back(o);
```
```suggestion
      nlohmann::json& o = writer.emplace_back();
      if (!InternalWrite(o, *itr))
        return false;
```

neo1973 commented on xbmc/utils/JSONVariantWriter.cpp:
```diff
@@ -10,83 +10,84 @@
     {
-      if (!writer.Key(itr->first.c_str()) ||
-        !InternalWrite(writer, itr->second))
+      nlohmann::json o;
+      if (InternalWrite(o, itr->second))
+        writer[itr->first] = std::move(o);
+      else
         return false;
```
```suggestion
      nlohmann::json& o = writer[itr->first];
      if (!InternalWrite(o, itr->second))
        return false;
```

neo1973 reviewed (commented):
Those two changes reduce the time to ~180ms for my data set.

sundermann commented:
I've applied those two changes and also changed the return type to `void` as the API uses exceptions to indicate failures and we catch these in the `Write` function.

neo1973 commented:
Any idea why the FreeBSD builder is unhappy?

sundermann commented:
It needs the dependency installed because FreeBSD doesn't do an internal build

sundermann commented:
Updated depends nlohmann-json to 3.12.0 which includes cmake 4.0 support

fuzzard commented:
Can you apply the following fixup https://github.com/fuzzard/xbmc/commit/60b5a39e9f121372f1679e924e41b0a8bf27adb5

older nlohmann_json configs have a target called nlohmann_json, so the build name has to be overridden to avoid potential clashes. Theres also been some recent changes regarding the spec search stuff

Can you also alter the ARCHIVE name in the VERSION file

```
ARCHIVE=$(LIBNAME)-$(VERSION).tar.xz
``` 

We need to be able to distinguish version on the mirror, and we cant just override an older archive with the tarball name they use for their releases (json.tar.xz)

The file ``json-3.12.0.tar.xz`` has been put on the mirrors now

sundermann commented:
@fuzzard Thanks for your build system changes and uploading the tarball. I've applied the fixup and removed the `BASE_URL` change

neo1973 commented:
Jenkins build this please

neo1973 commented:
Jenkins build this please

neo1973 reviewed (approved):
Jenkins is now green as well :+1:

sundermann commented on cmake/modules/FindNlohmannJSON.cmake:
```diff
@@ -0,0 +1,85 @@
+                             ${${CORE_PLATFORM_NAME_LC}_SEARCH_CONFIG})
+
+  # Check for existing NLOHMANNJSON. If version >= NLOHMANNJSON-VERSION file version, dont build
+  # A corner case, but if a linux/freebsd user WANTS to build internal tinyxml2, build anyway
```
Good catch, that was already a typo in `FindRapidjson.cmake` and carried over.
